### ISSUE TITLE
[Feature][Master] Optimize complement task's date

### ISSUE THREAD

**Issue by SbloodyS** [apache/dolphinscheduler#7576]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

![](https://ftp.bmp.ovh/imgs/2021/12/3b6c71080f2d3cf3.png)

Current complement task's date is left closed and right open time interval (startDate <= N < endDate).

From a user's point of view, what you see is what you get.In order that users do not need to transform understanding.I suggest changing left closed and right open time interval to left closed and right closed time interval.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by geosmart** [comment]
+1

**Comment by janeHe13** [comment]
After parallel complement, the time range of parallel complement is incorrect
The selection time is 2022.1.1 ~ 2022.1.3, but only one day of 2022.1.2 data is supplemented
![image](https://user-images.githubusercontent.com/95271106/149259243-577a800f-986b-4817-9474-9f307e9f8557.png)
![image](https://user-images.githubusercontent.com/95271106/149259304-0f21c468-0ef6-406f-a59e-45b145ef656d.png)

![image](https://user-images.githubusercontent.com/95271106/149259116-aa91493b-a9d2-4016-88ae-c8fc08ad73f9.png)

**Comment by caishunfeng** [comment]
> After parallel complement, the time range of parallel complement is incorrect The selection time is 2022.1.1 ~ 2022.1.3, but only one day of 2022.1.2 data is supplemented ![image](https://user-images.githubusercontent.com/95271106/149259243-577a800f-986b-4817-9474-9f307e9f8557.png) ![image](https://user-images.githubusercontent.com/95271106/149259304-0f21c468-0ef6-406f-a59e-45b145ef656d.png)
> 
> ![image](https://user-images.githubusercontent.com/95271106/149259116-aa91493b-a9d2-4016-88ae-c8fc08ad73f9.png)

No problem of complement task running. So thie issue will be closed again.
@janeHe13 you can create a new issue of the UI bug.

**Comment by zhongjiajie** [comment]
Why we should change strategy to this? IMO, I think we should let users not select the same time in UI, and check is not same datetime in backend. I think using strategy `left closed and right open` is good.

**Comment by SbloodyS** [comment]
In order let users do not need to transform understanding. ```left closed and right open``` is a hidden condition. As a user, I often forget this condition when making complement tasks, resulting in missing data.@zhongjiajie

**Comment by zhongjiajie** [comment]
And how about the date range in "timing", it is `left closed and right open` or is `left closed and right closed `?
<img width="1908" alt="image" src="https://user-images.githubusercontent.com/15820530/149895104-27f3c6e4-a2c7-4516-a64b-ee175d748b1b.png">

**Comment by zhongjiajie** [comment]
How about this other daterange fliters in DolphinScheduler? such as "project list"
<img width="1910" alt="image" src="https://user-images.githubusercontent.com/15820530/149895826-2d555733-68e8-4f75-a373-0f2a585da1bd.png">

**Comment by SbloodyS** [comment]
```java
org.apache.dolphinscheduler.service.quartz.cron.CronUtils#getSelfFireDateList(java.util.Date, java.util.Date, java.util.List<org.apache.dolphinscheduler.dao.entity.Schedule>)
```
This method is only used in complement tasks. Then the rest of the daterange filters remains the same.@zhongjiajie

**Comment by zhongjiajie** [comment]
I think is a little odd here. We have different behavior in our date range filters.

**Comment by SbloodyS** [comment]
> I think is a little odd here. We have different behavior in our date range filters.

In order to make it easy for users to understand.I think we should revise other time ranges to be consistent.Just like the same as complement data.

eg:
2022.01.01~2022.01.10

The date range currently selected in the UI:
2022.01.01~2022.01.11

After:
2022.01.01~2022.01.10

**Comment by zhongjiajie** [comment]
> > I think is a little odd here. We have different behavior in our date range filters.
> 
> In order to make it easy for users to understand.I think we should revise other time ranges to be consistent.Just like the same as complement data.
> 
> eg: 2022.01.01~2022.01.10
> 
> The date range currently selected in the UI: 2022.01.01~2022.01.11
> 
> After: 2022.01.01~2022.01.10

Yeah, but indeed, I thinks we should keep date range filter to `left closed and right open`. It make more sence for me. Maybe we should raise an mail thread to discuss it @SbloodyS . **BUT** no matter what conclusion we come to, I think we  should keep then same behavior in all our date range filter, to make user in the logic

**Comment by SbloodyS** [comment]
> > > I think is a little odd here. We have different behavior in our date range filters.
> > 
> > 
> > In order to make it easy for users to understand.I think we should revise other time ranges to be consistent.Just like the same as complement data.
> > eg: 2022.01.01~2022.01.10
> > The date range currently selected in the UI: 2022.01.01~2022.01.11
> > After: 2022.01.01~2022.01.10
> 
> Yeah, but indeed, I thinks we should keep date range filter to `left closed and right open`. It make more sence for me. Maybe we should raise an mail thread to discuss it @SbloodyS . **BUT** no matter what conclusion we come to, I think we should keep then same behavior in all our date range filter, to make user in the logic

Okay. I'll raise an mail thread to discuss it tonight or tomorrow.
